core/drop_test.go

```diff
@@ -0,0 +1,255 @@
+package core
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/0xPolygon/go-ibft/messages/proto"
+	"github.com/stretchr/testify/assert"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestDropMaxFaulty(t *testing.T) {
+	cluster := newCluster(
+		5,
+		func(c *cluster) {
+			for _, node := range c.nodes {
+				node.core = NewIBFT(
+					mockLogger{},
+					&mockBackend{
+						isValidBlockFn:         isValidProposal,
+						isValidProposalHashFn:  isValidProposalHash,
+						isValidSenderFn:        nil,
+						isValidCommittedSealFn: nil,
+						isProposerFn:           c.isProposer,
+
+						idFn:                 node.addr,
+						quorumFn:             quorum,
+						maximumFaultyNodesFn: c.maxFaulty,
+
+						buildProposalFn:           buildValidProposal,
+						buildPrePrepareMessageFn:  node.buildPrePrepare,
+						buildPrepareMessageFn:     node.buildPrepare,
+						buildCommitMessageFn:      node.buildCommit,
+						buildRoundChangeMessageFn: node.buildRoundChange,
+						insertBlockFn:             nil,
+					},
+
+					&mockTransport{multicastFn: c.gossip},
+				)
+			}
+		},
+	)
+
+	if err := cluster.progressToHeight(5*time.Second, 5); err != nil {
+		t.Fatal("cannot progress to height: err=", err)
+	}
+
+	cluster.stopN(int(cluster.maxFaulty()) + 1)
```

Comment:
Can you use random nodes there?

Reply:
I didn't see any need for that in this particular case scenario. All the nodes are identical except for the [offline behavior](https://github.com/0xPolygon/go-ibft/blob/b03b4031826dad456fbe2856ac59abefa3b1ac5d/core/drop_test.go#L173), so having a random distribution does not affect the test flow.